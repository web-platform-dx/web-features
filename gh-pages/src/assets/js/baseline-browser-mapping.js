import bcdBrowsers from 'https://cdn.jsdelivr.net/npm/@mdn/browser-compat-data' with { type: 'json' };
import otherBrowsers from "https://cdn.jsdelivr.net/npm/baseline-browser-mapping/dist/data/downstream-browsers.json" with {type: "json"}
import webFeatures from "https://cdn.jsdelivr.net/npm/web-features/data.json" with {type: "json"}
const features = webFeatures.features;
const bcdCoreBrowserNames = [
  "chrome",
  "chrome_android",
  "edge",
  "firefox",
  "firefox_android",
  "safari",
  "safari_ios",
];
const coreBrowserData = Object.entries(bcdBrowsers.browsers).filter(([browserName]) => bcdCoreBrowserNames.includes(browserName));
const bcdDownstreamBrowserNames = [
  "webview_android",
  "samsunginternet_android",
  "opera_android",
  "opera",
];
const downstreamBrowserData = [
  ...Object.entries(bcdBrowsers.browsers).filter(([browserName]) => bcdDownstreamBrowserNames.includes(browserName)),
  ...Object.entries(otherBrowsers.browsers),
];
const acceptableStatuses = [
  "current",
  "esr",
  "retired",
  "unknown",
  "beta",
  "nightly",
];
let suppressPre2015Warning = false;
const stripLTEPrefix = (str) => {
  if (!str) {
    return str;
  }
  if (!str.startsWith("â‰¤")) {
    return str;
  }
  return str.slice(1);
};
const compareVersions = (incomingVersionString, previousVersionString) => {
  if (incomingVersionString === previousVersionString) {
    return 0;
  }
  let [incomingVersionStringMajor, incomingVersionStringMinor] = incomingVersionString.split(".");
  let [previousVersionStringMajor, previousVersionStringMinor] = previousVersionString.split(".");
  if (!incomingVersionStringMajor || !previousVersionStringMajor) {
    throw new Error("One of these version strings is broken: " +
      incomingVersionString +
      " or " +
      previousVersionString +
      "");
  }
  if (parseInt(incomingVersionStringMajor) > parseInt(previousVersionStringMajor)) {
    return 1;
  }
  if (incomingVersionStringMinor) {
    if (parseInt(incomingVersionStringMajor) ==
      parseInt(previousVersionStringMajor) &&
      (!previousVersionStringMinor ||
        parseInt(incomingVersionStringMinor) >
        parseInt(previousVersionStringMinor))) {
      return 1;
    }
  }
  return -1;
};
const getCompatibleFeaturesByDate = (date) => {
  return Object.entries(features)
    .filter(([feature_id, feature]) => feature.status.baseline_low_date &&
      new Date(feature.status.baseline_low_date) <= date)
    .map(([feature_id, feature]) => {
      return {
        id: feature_id,
        name: feature.name,
        baseline_low_date: feature.status.baseline_low_date,
        support: feature.status.support,
      };
    });
};
const getMinimumVersionsFromFeatures = (features) => {
  let minimumVersions = {};
  Object.entries(coreBrowserData).forEach(([, browserData]) => {
    minimumVersions[browserData[0]] = {
      browser: browserData[0],
      version: "0",
      release_date: "",
    };
  });
  features.forEach((feature) => {
    Object.entries(feature.support).forEach((browser) => {
      const browserName = browser[0];
      const version = stripLTEPrefix(browser[1]);
      if (minimumVersions[browserName] &&
        compareVersions(version, stripLTEPrefix(minimumVersions[browserName].version)) === 1) {
        minimumVersions[browserName] = {
          browser: browserName,
          version: version,
          release_date: feature.baseline_low_date,
        };
      }
    });
  });
  return Object.values(minimumVersions);
};
const getSubsequentVersions = (minimumVersions) => {
  let subsequentVersions = [];
  minimumVersions.forEach((minimumVersion) => {
    let bcdBrowser = coreBrowserData.find((bcdBrowser) => bcdBrowser[0] === minimumVersion.browser);
    if (bcdBrowser) {
      let sortedVersions = Object.entries(bcdBrowser[1].releases)
        .filter(([, versionData]) => {
          if (!acceptableStatuses.includes(versionData.status)) {
            return false;
          }
          return true;
        })
        .sort((a, b) => {
          return compareVersions(a[0], b[0]);
        });
      sortedVersions.forEach(([version, versionData]) => {
        if (!acceptableStatuses.includes(versionData.status)) {
          return false;
        }
        if (compareVersions(version, minimumVersion.version) === 1) {
          subsequentVersions.push({
            browser: minimumVersion.browser,
            version: version,
            release_date: versionData.release_date
              ? versionData.release_date
              : "unknown",
          });
          return true;
        }
        return false;
      });
    }
  });
  return subsequentVersions;
};
const getCoreVersionsByDate = (date, listAllCompatibleVersions = false) => {
  if (date.getFullYear() < 2015 && !suppressPre2015Warning) {
    console.warn(new Error("There are no browser versions compatible with Baseline before 2015.  You may receive unexpected results."));
  }
  if (date.getFullYear() < 2002) {
    throw new Error("None of the browsers in the core set were released before 2002.  Please use a date after 2002.");
  }
  if (date.getFullYear() > new Date().getFullYear()) {
    throw new Error("There are no browser versions compatible with Baseline in the future");
  }
  const compatibleFeatures = getCompatibleFeaturesByDate(date);
  const minimumVersions = getMinimumVersionsFromFeatures(compatibleFeatures);
  if (!listAllCompatibleVersions) {
    return minimumVersions;
  }
  else {
    return [...minimumVersions, ...getSubsequentVersions(minimumVersions)].sort((a, b) => {
      if (a.browser < b.browser) {
        return -1;
      }
      else if (a.browser > b.browser) {
        return 1;
      }
      else {
        return compareVersions(a.version, b.version);
      }
    });
  }
};
const getDownstreamBrowsers = (inputArray = [], listAllCompatibleVersions = true) => {
  let minimumChromeVersion = undefined;
  if (inputArray && inputArray.length > 0) {
    minimumChromeVersion = inputArray
      .filter((browser) => browser.browser === "chrome")
      .sort((a, b) => {
        return compareVersions(a.version, b.version);
      })[0]?.version;
  }
  if (!minimumChromeVersion) {
    throw new Error("There are no browser versions compatible with Baseline before Chrome");
  }
  let downstreamArray = new Array();
  downstreamBrowserData.forEach(([browserName, browserData]) => {
    if (!browserData.releases)
      return;
    let sortedAndFilteredVersions = Object.entries(browserData.releases)
      .filter(([, versionData]) => {
        if (!versionData.engine) {
          return false;
        }
        if (versionData.engine != "Blink") {
          return false;
        }
        if (versionData.engine_version &&
          parseInt(versionData.engine_version) < parseInt(minimumChromeVersion)) {
          return false;
        }
        return true;
      })
      .sort((a, b) => {
        return compareVersions(a[0], b[0]);
      });
    for (let i = 0; i < sortedAndFilteredVersions.length; i++) {
      const versionEntry = sortedAndFilteredVersions[i];
      if (versionEntry) {
        const [versionNumber, versionData] = versionEntry;
        let outputArray = {
          browser: browserName,
          version: versionNumber,
          release_date: versionData.release_date ?? "unknown",
        };
        if (versionData.engine && versionData.engine_version) {
          outputArray.engine = versionData.engine;
          outputArray.engine_version = versionData.engine_version;
        }
        downstreamArray.push(outputArray);
        if (!listAllCompatibleVersions) {
          break;
        }
      }
    }
  });
  return downstreamArray;
};
/**
 * Returns browser versions compatible with specified Baseline targets.
 * Defaults to returning the minimum versions of the core browser set that support Baseline Widely available.
 * Takes an optional configuraation object `Object` with four optional properties:
 * - `listAllCompatibleVersions`: `false` (default) or `false`
 * - `includeDownstreamBrowsers`: `false` (default) or `false`
 * - `widelyAvailableOnDate`: date in format `YYYY-MM-DD`
 * - `targetYear`: year in format `YYYY`
 */
export function getCompatibleVersions(userOptions) {
  let incomingOptions = userOptions ?? {};
  let options = {
    listAllCompatibleVersions: incomingOptions.listAllCompatibleVersions ?? false,
    includeDownstreamBrowsers: incomingOptions.includeDownstreamBrowsers ?? false,
    widelyAvailableOnDate: incomingOptions.widelyAvailableOnDate ?? undefined,
    targetYear: incomingOptions.targetYear ?? undefined,
  };
  let targetDate = new Date();
  if (!options.widelyAvailableOnDate && !options.targetYear) {
    targetDate = new Date();
  }
  else if (options.targetYear && options.widelyAvailableOnDate) {
    console.log(new Error("You cannot use targetYear and widelyAvailableOnDate at the same time.  Please remove one of these options and try again."));
    process.exit(1);
  }
  else if (options.widelyAvailableOnDate) {
    targetDate = new Date(options.widelyAvailableOnDate);
  }
  else if (options.targetYear) {
    targetDate = new Date(`${options.targetYear}-12-31`);
  }
  // Sets a cutoff date for feature interoperability 30 months before the stated date
  if (options.widelyAvailableOnDate || options.targetYear === undefined) {
    targetDate.setMonth(targetDate.getMonth() - 30);
  }
  let coreBrowserArray = getCoreVersionsByDate(targetDate, options.listAllCompatibleVersions);
  if (options.includeDownstreamBrowsers === false) {
    return coreBrowserArray;
  }
  else {
    return [
      ...coreBrowserArray,
      ...getDownstreamBrowsers(coreBrowserArray, options.listAllCompatibleVersions),
    ];
  }
}
/**
 * Returns all browser versions known to this module with their level of Baseline support either as an `Array` or a `String` CSV.
 * Takes an object as an argument with two optional properties:
 * - `includeDownstreamBrowsers`: `true` (default) or `false`
 * - `outputFormat`: `array` (default), `object` or `csv`
 */
export function getAllVersions(userOptions) {
  suppressPre2015Warning = true;
  let incomingOptions = userOptions ?? {};
  let options = {
    outputFormat: incomingOptions.outputFormat ?? "array",
    includeDownstreamBrowsers: incomingOptions.includeDownstreamBrowsers ?? false,
    useSupports: incomingOptions.useSupports ?? false,
  };
  let nextYear = new Date().getFullYear() + 1;
  const yearArray = [...Array(nextYear).keys()].slice(2002);
  const yearMinimumVersions = {};
  yearArray.forEach((year) => {
    yearMinimumVersions[year] = {};
    getCompatibleVersions({ targetYear: year }).forEach((version) => {
      if (yearMinimumVersions[year])
        yearMinimumVersions[year][version.browser] = version;
    });
  });
  const waMinimumVersions = getCompatibleVersions({});
  const waObject = {};
  waMinimumVersions.forEach((version) => {
    waObject[version.browser] = version;
  });
  const thirtyMonthsFromToday = new Date();
  thirtyMonthsFromToday.setMonth(thirtyMonthsFromToday.getMonth() + 30);
  const naMinimumVersions = getCompatibleVersions({
    widelyAvailableOnDate: thirtyMonthsFromToday.toISOString().slice(0, 10),
  });
  const naObject = {};
  naMinimumVersions.forEach((version) => {
    naObject[version.browser] = version;
  });
  const allVersions = getCompatibleVersions({
    targetYear: 2002,
    listAllCompatibleVersions: true,
  });
  const outputArray = new Array();
  bcdCoreBrowserNames.forEach((browserName) => {
    let thisBrowserAllVersions = allVersions
      .filter((version) => version.browser == browserName)
      .sort((a, b) => {
        return compareVersions(a.version, b.version);
      });
    let waVersion = waObject[browserName]?.version ?? "0";
    let naVersion = naObject[browserName]?.version ?? "0";
    yearArray.forEach((year) => {
      if (yearMinimumVersions[year]) {
        let minBrowserVersionInfo = yearMinimumVersions[year][browserName] ?? {
          version: "0",
        };
        let minBrowserVersion = minBrowserVersionInfo.version;
        let sliceIndex = thisBrowserAllVersions.findIndex((element) => compareVersions(element.version, minBrowserVersion) === 0);
        let subArray = year === nextYear - 1
          ? thisBrowserAllVersions
          : thisBrowserAllVersions.slice(0, sliceIndex);
        subArray.forEach((version) => {
          let isWaCcompatible = compareVersions(version.version, waVersion) >= 0 ? true : false;
          let isNaCompatible = compareVersions(version.version, naVersion) >= 0 ? true : false;
          let versionToPush = {
            ...version,
            year: year <= 2015 ? "pre_baseline" : year - 1,
          };
          if (options.useSupports) {
            if (isWaCcompatible)
              versionToPush.supports = "widely";
            if (isNaCompatible)
              versionToPush.supports = "newly";
          }
          else {
            versionToPush = {
              ...versionToPush,
              wa_compatible: isWaCcompatible,
            };
          }
          outputArray.push(versionToPush);
        });
        thisBrowserAllVersions = thisBrowserAllVersions.slice(sliceIndex, thisBrowserAllVersions.length);
      }
    });
  });
  if (options.includeDownstreamBrowsers) {
    let downstreamBrowsers = getDownstreamBrowsers(outputArray);
    downstreamBrowsers.forEach((version) => {
      let correspondingChromiumVersion = outputArray.find((upstreamVersion) => upstreamVersion.browser === "chrome" &&
        upstreamVersion.version === version.engine_version);
      if (correspondingChromiumVersion) {
        if (options.useSupports) {
          outputArray.push({
            ...version,
            year: correspondingChromiumVersion.year,
            supports: correspondingChromiumVersion.supports,
          });
        }
        else {
          outputArray.push({
            ...version,
            year: correspondingChromiumVersion.year,
            wa_compatible: correspondingChromiumVersion.wa_compatible,
          });
        }
      }
    });
  }
  outputArray.sort((a, b) => {
    // Sort by year: "pre_baseline" first, then numerical year in ascending order
    if (a.year === "pre_baseline" && b.year !== "pre_baseline") {
      return -1;
    }
    if (b.year === "pre_baseline" && a.year !== "pre_baseline") {
      return 1;
    }
    if (a.year !== "pre_baseline" && b.year !== "pre_baseline") {
      if (a.year < b.year) {
        return -1;
      }
      if (a.year > b.year) {
        return 1;
      }
    }
    // Sort by browser alphabetically
    if (a.browser < b.browser) {
      return -1;
    }
    if (a.browser > b.browser) {
      return 1;
    }
    // Sort by version using compareVersions
    return compareVersions(a.version, b.version);
  });
  if (options.outputFormat === "object") {
    const outputObject = {};
    outputArray.forEach((version) => {
      if (!outputObject[version.browser]) {
        outputObject[version.browser] = {};
      }
      let versionToAdd = {
        year: version.year,
        release_date: version.release_date,
        engine: version.engine,
        engine_version: version.engine_version,
      };
      if (options.useSupports) {
        //@ts-ignore
        outputObject[version.browser][version.version] = version.supports
          ? { ...versionToAdd, supports: version.supports }
          : versionToAdd;
      }
      else {
        //@ts-ignores
        outputObject[version.browser][version.version] = {
          ...versionToAdd,
          wa_compatible: version.wa_compatible,
        };
      }
    });
    return outputObject ?? {};
  }
  if (options.outputFormat === "csv") {
    let outputString = `"browser","version","year",` +
      `"${options.useSupports ? "supports" : "wa_compatible"}",` +
      `"release_date","engine","engine_version"`;
    outputArray.forEach((version) => {
      let outputs = {
        browser: version.browser,
        version: version.version,
        year: version.year,
        release_date: version.release_date ?? "NULL",
        engine: version.engine ?? "NULL",
        engine_version: version.engine_version ?? "NULL",
      };
      outputs = options.useSupports
        ? { ...outputs, supports: version.supports ?? "" }
        : { ...outputs, wa_compatible: version.wa_compatible };
      outputString +=
        `\n"${outputs.browser}","` +
        `${outputs.version}","` +
        `${outputs.year}","` +
        `${options.useSupports ? outputs.supports : outputs.wa_compatible}","` +
        `${outputs.release_date}","` +
        `${outputs.engine}","` +
        `${outputs.engine_version}"`;
    });
    return outputString;
  }
  return outputArray;
}
