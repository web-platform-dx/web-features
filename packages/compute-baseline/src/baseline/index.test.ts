import assert from "node:assert/strict";

import { Temporal } from "@js-temporal/polyfill";
import * as chai from "chai";
import chaiJestSnapshot from "chai-jest-snapshot";

import { browser } from "../browser-compat-data/index.js";
import { computeBaseline, getStatus, keystoneDateToStatus } from "./index.js";

chai.use(chaiJestSnapshot);

describe("getStatus", function () {
  before(function () {
    chaiJestSnapshot.resetSnapshotRegistry();
  });

  beforeEach(function () {
    chaiJestSnapshot.configureUsingMochaContext(this);
  });

  it("returns a status", function () {
    const result = getStatus("fetch", "api.Response.json");
    assert.equal(result.baseline, "high");
    chai.expect(result).to.matchSnapshot();
  });
});

describe("computeBaseline", function () {
  before(function () {
    chaiJestSnapshot.resetSnapshotRegistry();
  });

  beforeEach(function () {
    chaiJestSnapshot.configureUsingMochaContext(this);
  });

  it("returns something sensible for the most complex features", function () {
    // These are some of the most "complex" BCD features, given approximately by:
    // const complexity = JSON.stringify(data.__compat, undefined, 2).split("\n").length
    const result = Object.fromEntries(
      [
        "api.Document.visibilitychange_event",
        "css.types.basic-shape.path",
        "api.DOMMatrix.DOMMatrix",
      ].map((key) => [
        key,
        computeBaseline({ compatKeys: [key], checkAncestors: false }),
      ]),
    );
    assert.equal(
      result["api.Document.visibilitychange_event"]?.baseline,
      "high",
    );
    assert.equal(result["css.types.basic-shape.path"]?.baseline, false);
    assert.equal(result["api.DOMMatrix.DOMMatrix"]?.baseline, "high");
    chai.expect(result).to.matchSnapshot();
  });

  it("returns something sensible for the least complex features", function () {
    // These are some of the least "complex" BCD features
    const result = Object.fromEntries(
      [
        "css.properties.counter-reset.reset_does_not_affect_siblings",
        "html.global_attributes.exportparts",
        "html.elements.form.target",
      ].map((key) => [
        key,
        computeBaseline({ compatKeys: [key], checkAncestors: false }),
      ]),
    );

    assert.equal(
      result["css.properties.counter-reset.reset_does_not_affect_siblings"]
        ?.baseline,
      false,
    );
    assert.equal(
      result["html.global_attributes.exportparts"]?.baseline,
      "high",
    );
    assert.equal(result["html.elements.form.target"]?.baseline, "high");
    chai.expect(result).to.matchSnapshot();
  });

  it("returns a result for a feature", function () {
    const result = computeBaseline({
      compatKeys: ["css.properties.border-color"],
      checkAncestors: false,
    });
    assert.equal(result.baseline, "high");
    assert.equal(result.baseline_low_date, "2015-07-29"); // The first release of Edge, the youngest release in consideration
    assert.equal(result.baseline_high_date, "2018-01-29"); // 30 months later
  });

  it("finds discrepancies with ancestors (checkAncestors)", function () {
    const result = computeBaseline({
      compatKeys: ["api.Notification.body"],
      checkAncestors: false,
    });
    const resultExplicit = computeBaseline({
      compatKeys: ["api.Notification", "api.Notification.body"],
      checkAncestors: false,
    });
    const resultWithAncestors = computeBaseline({
      compatKeys: ["api.Notification.body"],
      checkAncestors: true,
    });

    assert.equal(resultExplicit.toJSON(), resultWithAncestors.toJSON());
    assert.notEqual(result.toJSON(), resultWithAncestors.toJSON());

    assert.notEqual(result.baseline, resultWithAncestors.baseline);
    assert.notEqual(
      result.baseline_low_date?.toString(),
      resultWithAncestors.baseline_low_date?.toString(),
    );

    chai.expect(result).to.matchSnapshot();
    chai.expect(resultExplicit).to.matchSnapshot();
    chai.expect(resultWithAncestors).to.matchSnapshot();
  });

  it("surfaces version ranges from the underlying compat data", function () {
    const singleKey = computeBaseline({
      compatKeys: ["css.properties.align-self"],
    });
    const singleKeySafari = singleKey.support.get(browser("safari"));
    assert(singleKeySafari);
    assert.equal(singleKeySafari.release.version, "9");
    assert.equal(singleKeySafari.text, "9");
    assert.equal(JSON.parse(singleKey.toJSON()).support.safari, "9");

    const multiKey = computeBaseline({
      compatKeys: [
        "css.properties.align-self",
        "css.properties.align-self.flex_context.baseline",
      ],
      checkAncestors: false,
    });
    const multiKeySafari = multiKey.support.get(browser("safari"));
    assert(multiKeySafari);
    assert.equal(multiKeySafari.release.version, "9");
    assert.equal(multiKeySafari.text, "9");
    assert.equal(JSON.parse(multiKey.toJSON()).support.safari, "9");
  });

  it("calculates ranged dates from underlying compat data ranged versions", function () {
    const result = computeBaseline({
      compatKeys: ["api.FileSystem"],
    });
    assert.equal(JSON.parse(result.toJSON()).support.edge, "≤18");
    assert(JSON.parse(result.toJSON()).baseline_low_date.startsWith("≤"));
  });

  it("disregards support that's been removed", function () {
    const result = computeBaseline({
      compatKeys: ["api.AudioTrack"],
      checkAncestors: false,
    });
    chai.expect(result).to.matchSnapshot();
    assert.notEqual(Boolean(result.baseline), true);
  });

  it("rejects deprecated", function () {
    const actual = computeBaseline({
      compatKeys: ["javascript.statements.with"],
      checkAncestors: false,
    });
    assert.equal(actual.baseline, false);
  });
});

describe("keystoneDateToStatus()", function () {
  it('returns "low" for date 1 year before cutoff date', function () {
    const status = keystoneDateToStatus(
      "2020-01-01",
      Temporal.PlainDate.from("2021-01-01"),
      false,
    );
    assert.equal(status.baseline, "low");
    assert.equal(status.baseline_low_date, "2020-01-01");
    assert.equal(status.baseline_high_date, null);
  });

  it('returns "high" for date 3 years before cutoff date', function () {
    const status = keystoneDateToStatus(
      "2020-01-01",
      Temporal.PlainDate.from("2024-01-01"),
      false,
    );
    assert.equal(status.baseline, "high");
    assert.equal(status.baseline_low_date, "2020-01-01");
    assert.equal(status.baseline_high_date, "2022-07-01");
  });

  it("returns false for null dates", function () {
    const status = keystoneDateToStatus(
      null,
      Temporal.PlainDate.from("2020-01-01"),
      false,
    );
    assert.equal(status.baseline, false);
    assert.equal(status.baseline_low_date, null);
    assert.equal(status.baseline_high_date, null);
  });

  it("returns false for discouraged (deprecated, obsolete, etc.) features", function () {
    const status = keystoneDateToStatus(
      "2020-01-01",
      Temporal.PlainDate.from("2020-01-01"),
      true,
    );
    assert.equal(status.baseline, false);
    assert.equal(status.baseline_low_date, null);
    assert.equal(status.baseline_high_date, null);
  });

  it("preserves ranges where they exist", function () {
    const status = keystoneDateToStatus(
      "≤2020-01-01",
      Temporal.PlainDate.from("2024-01-01"),
      false,
    );
    assert.equal(status.baseline, "high");
    assert.equal(status.baseline_low_date, "≤2020-01-01");
    assert.equal(status.baseline_high_date, "≤2022-07-01");
  });
});
