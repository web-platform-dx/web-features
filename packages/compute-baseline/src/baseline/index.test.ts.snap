// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`computeBaseline disregards support that's been removed 1`] = `
"{
  \\"baseline\\": false,
  \\"support\\": {
    \\"safari\\": \\"8\\",
    \\"safari_ios\\": \\"8\\"
  }
}"
`;

exports[`computeBaseline finds discrepancies with ancestors (checkAncestors) 1`] = `
"{
  \\"baseline\\": \\"low\\",
  \\"baseline_low_date\\": \\"2023-03-27\\",
  \\"support\\": {
    \\"chrome\\": \\"69\\",
    \\"chrome_android\\": \\"69\\",
    \\"edge\\": \\"79\\",
    \\"firefox\\": \\"64\\",
    \\"firefox_android\\": \\"64\\",
    \\"safari\\": \\"16.4\\",
    \\"safari_ios\\": \\"16.4\\"
  }
}"
`;

exports[`computeBaseline finds discrepancies with ancestors (checkAncestors) 2`] = `
"{
  \\"baseline\\": false,
  \\"support\\": {
    \\"chrome\\": \\"71\\",
    \\"chrome_android\\": \\"71\\",
    \\"edge\\": \\"79\\",
    \\"firefox\\": \\"64\\",
    \\"firefox_android\\": \\"64\\",
    \\"safari\\": \\"16.4\\"
  }
}"
`;

exports[`computeBaseline finds discrepancies with ancestors (checkAncestors) 3`] = `
"{
  \\"baseline\\": false,
  \\"support\\": {
    \\"chrome\\": \\"71\\",
    \\"chrome_android\\": \\"71\\",
    \\"edge\\": \\"79\\",
    \\"firefox\\": \\"64\\",
    \\"firefox_android\\": \\"64\\",
    \\"safari\\": \\"16.4\\"
  }
}"
`;

exports[`computeBaseline returns something sensible for the least complex features 1`] = `
Object {
  "css.properties.counter-reset.reset_does_not_affect_siblings": "{
  \\"baseline\\": false,
  \\"support\\": {
    \\"firefox\\": \\"82\\",
    \\"firefox_android\\": \\"82\\"
  }
}",
  "html.elements.form.target": "{
  \\"baseline\\": \\"high\\",
  \\"baseline_low_date\\": \\"2015-07-29\\",
  \\"baseline_high_date\\": \\"2018-01-29\\",
  \\"support\\": {
    \\"chrome\\": \\"1\\",
    \\"chrome_android\\": \\"18\\",
    \\"edge\\": \\"12\\",
    \\"firefox\\": \\"1\\",
    \\"firefox_android\\": \\"4\\",
    \\"safari\\": \\"3\\",
    \\"safari_ios\\": \\"2\\"
  }
}",
  "html.global_attributes.exportparts": "{
  \\"baseline\\": \\"high\\",
  \\"baseline_low_date\\": \\"2020-07-28\\",
  \\"baseline_high_date\\": \\"2023-01-28\\",
  \\"support\\": {
    \\"chrome\\": \\"73\\",
    \\"chrome_android\\": \\"73\\",
    \\"edge\\": \\"79\\",
    \\"firefox\\": \\"72\\",
    \\"firefox_android\\": \\"79\\",
    \\"safari\\": \\"13.1\\",
    \\"safari_ios\\": \\"13.4\\"
  }
}",
}
`;

exports[`computeBaseline returns something sensible for the most complex features 1`] = `
Object {
  "api.DOMMatrix.DOMMatrix": "{
  \\"baseline\\": \\"high\\",
  \\"baseline_low_date\\": \\"2020-01-15\\",
  \\"baseline_high_date\\": \\"2022-07-15\\",
  \\"support\\": {
    \\"chrome\\": \\"61\\",
    \\"chrome_android\\": \\"61\\",
    \\"edge\\": \\"79\\",
    \\"firefox\\": \\"33\\",
    \\"firefox_android\\": \\"33\\",
    \\"safari\\": \\"11\\",
    \\"safari_ios\\": \\"11\\"
  }
}",
  "api.Document.visibilitychange_event": "{
  \\"baseline\\": \\"high\\",
  \\"baseline_low_date\\": \\"2021-04-26\\",
  \\"baseline_high_date\\": \\"2023-10-26\\",
  \\"support\\": {
    \\"chrome\\": \\"62\\",
    \\"chrome_android\\": \\"62\\",
    \\"edge\\": \\"18\\",
    \\"firefox\\": \\"56\\",
    \\"firefox_android\\": \\"56\\",
    \\"safari\\": \\"14.1\\",
    \\"safari_ios\\": \\"14.5\\"
  }
}",
  "css.types.basic-shape.path": "{
  \\"baseline\\": \\"high\\",
  \\"baseline_low_date\\": \\"2020-07-28\\",
  \\"baseline_high_date\\": \\"2023-01-28\\",
  \\"support\\": {
    \\"chrome\\": \\"52\\",
    \\"chrome_android\\": \\"52\\",
    \\"edge\\": \\"79\\",
    \\"firefox\\": \\"72\\",
    \\"firefox_android\\": \\"79\\",
    \\"safari\\": \\"13.1\\",
    \\"safari_ios\\": \\"13\\"
  }
}",
  "css.types.image.cross-fade": "{
  \\"baseline\\": false,
  \\"support\\": {
    \\"safari\\": \\"10\\",
    \\"safari_ios\\": \\"9.3\\"
  }
}",
}
`;

exports[`getStatus returns a status 1`] = `
Object {
  "baseline": "high",
  "baseline_high_date": "2019-09-27",
  "baseline_low_date": "2017-03-27",
  "support": Object {
    "chrome": "42",
    "chrome_android": "42",
    "edge": "14",
    "firefox": "39",
    "firefox_android": "39",
    "safari": "10.1",
    "safari_ios": "10.3",
  },
}
`;
